From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 1fexd <PGFvy85nC@protonmail.com>
Date: Wed, 17 May 2023 17:50:35 +0200
Subject: [PATCH] chore: Display serialized Material You color using serializer


diff --git a/app/src/main/kotlin/dev/patrickgold/florisboard/app/settings/theme/ThemeTranslations.kt b/app/src/main/kotlin/dev/patrickgold/florisboard/app/settings/theme/ThemeTranslations.kt
index 46260f9c..a8ff0458 100644
--- a/app/src/main/kotlin/dev/patrickgold/florisboard/app/settings/theme/ThemeTranslations.kt
+++ b/app/src/main/kotlin/dev/patrickgold/florisboard/app/settings/theme/ThemeTranslations.kt
@@ -19,7 +19,6 @@ package dev.patrickgold.florisboard.app.settings.theme
 import androidx.compose.runtime.Composable
 import androidx.compose.runtime.remember
 import androidx.compose.ui.graphics.Color
-import androidx.compose.ui.platform.LocalContext
 import dev.patrickgold.florisboard.R
 import dev.patrickgold.florisboard.ime.theme.FlorisImeUi
 import dev.patrickgold.florisboard.lib.UnicodeCtrlChar
@@ -37,7 +36,6 @@ import dev.patrickgold.florisboard.lib.snygg.value.SnyggExplicitInheritValue
 import dev.patrickgold.florisboard.lib.snygg.value.SnyggImplicitInheritValue
 import dev.patrickgold.florisboard.lib.snygg.value.SnyggMaterialYouDarkColorValue
 import dev.patrickgold.florisboard.lib.snygg.value.SnyggMaterialYouLightColorValue
-import dev.patrickgold.florisboard.lib.snygg.value.SnyggMaterialYouValue
 import dev.patrickgold.florisboard.lib.snygg.value.SnyggPercentageSizeValue
 import dev.patrickgold.florisboard.lib.snygg.value.SnyggRectangleShapeValue
 import dev.patrickgold.florisboard.lib.snygg.value.SnyggRoundedCornerDpShapeValue
@@ -160,14 +158,10 @@ internal fun translatePropertyValue(
     level: SnyggLevel,
     displayColorsAs: DisplayColorsAs,
 ): String {
-    val context = LocalContext.current
     return when (propertyValue) {
         is SnyggSolidColorValue -> remember(propertyValue.color, displayColorsAs) {
             buildColorString(propertyValue.color, displayColorsAs)
         }
-        is SnyggMaterialYouValue -> remember(propertyValue.colorName, displayColorsAs) {
-            buildColorString(propertyValue.loadColor(context), displayColorsAs)
-        }
         else -> when (level) {
             SnyggLevel.DEVELOPER -> null
             else -> when (propertyValue) {
-- 
2.41.0

