From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 1fexd <PGFvy85nC@protonmail.com>
Date: Wed, 17 May 2023 17:53:04 +0200
Subject: [PATCH] chore: Bum Compose Material3 dependency


diff --git a/gradle/libs.versions.toml b/gradle/libs.versions.toml
index 82248594..04d97aa5 100644
--- a/gradle/libs.versions.toml
+++ b/gradle/libs.versions.toml
@@ -6,7 +6,7 @@ androidx-activity = "1.5.1"
 androidx-autofill = "1.1.0"
 androidx-collection = "1.2.0"
 androidx-compose = "1.4.3"
-androidx-compose-material3 = "1.1.0-rc01"
+androidx-compose-material3 = "1.1.0"
 androidx-compose-compiler = "1.4.7"
 androidx-core = "1.10.0"
 androidx-core-splashscreen = "1.0.0"
-- 
2.41.0

